kubespawner/spawner.py

```diff
@@ -1757,6 +1783,82 @@ def start(self):
 
     _last_event = None
 
+    @gen.coroutine
+    def _make_create_pod_request(self, pod, request_timeout):
+        """
+        Make an HTTP request to create the given pod
+
+        Designed to be used with exponential_backoff, so returns
+        True / False on success / failure
+        """
+        try:
+            self.log.info(f"Attempting to create pod {pod.metadata.name}, with timeout {request_timeout}")
+            # Use tornado's timeout, _request_timeout seems unreliable?
+            yield gen.with_timeout(timedelta(seconds=request_timeout), self.asynchronize(
+                self.api.create_namespaced_pod,
+                self.namespace,
+                pod,
+            ))
+            return True
+        except gen.TimeoutError:
+            # Just try again
+            return False
+        except ApiException as e:
+            pod_name = pod.metadata.name
+            if e.status != 409:
+                # We only want to handle 409 conflict errors
+                self.log.exception("Failed for %s", pod.to_str())
+                raise
+            self.log.info(f'Found existing pod {pod_name}, attempting to kill')
+            # TODO: this should show up in events
+            yield self.stop(True)
+
+            self.log.info(f'Killed pod {pod_name}, will try starting singleuser pod again')
+            # We tell exponential_backoff to retry
+            return False
```

Comment:
I found the following [list of k8s api-server response status codes](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#http-status-codes) to be useful. For a _create resource_ request, the 409 status means the resource already exists.

I think this logic here is invalid or needs clarifications. Should we raise an error, return True / False, and/or take actions like deleting the pod etc in various situations?

Reply:
This logic is unchanged from before - if a pod is already created (i.e. not by us!), delete it and recreate it.

The very-short timeouts in this PR, though, make it a bit more likely that we:

1. create it successfully, but timeout is hit
2. next attempt fails with 409, but success was ours!
3. delete our first creation and try again

Reply:
Follow-up to clarify the last point: this case where we created the pod, API call raised but creation actually happened, and then we delete & retry was already technically a possibility that could occur. This PR adds one more situation where it can happen that can be triggered by load on *either* kubernetes or the Hub.